### DISCUSSION TITLE
Getting an IllegalStateException for non-existing class in Caffeine?

### DISCUSSION THREAD

**Discussion by Andre601** [ben-manes/caffeine#762] [Q&A] [answered]
I'm encountering a very special issue where I encounter an `IllegalStateException` with the reason `ch.andre601.advancedserverlist.core.depends.caffeine.cache.SSW` being provided.

Within the StackTrace is also a `Caused by: java.lang.ClassNotFoundException: ch.andre601.advancedserverlist.core.depends.caffeine.cache.SSW` which is weird as I relocate the Caffeine dependency and therefore the SSW class should also be present in the relocated pattern.

I'm using Caffeine as follows:
https://paste.helpch.at/nipucihaqi.java

The class itself is loaded within a void method in another class (Where it also gets the `T` declaration):
https://paste.helpch.at/idewovezon.java

Note that the above class is in a separate module that inherits/includes the core module which contains the Caffeine library.
The full project structure can be found here: https://github.com/Andre601/AdvancedServerList

I do shade in the library using relocations with the Maven Shade plugin.
My full pom.xml: https://paste.helpch.at/wojocafega.xml

I'm completely out of ideas what could be the cause here. The Cache worked fine before, until I decided to change the `FaviconHandler` class to `FaviconHandler<T>` to store `T` directly in a cache...
Does Caffeine not allow `T` and similar to be used? And if so, what are workarounds for it?

**Comment by ben-manes** [comment, marked as answer]
I suspect that some form of dead code elimination is pruning the class files. I see that you enabled `minimizeJar` which probably does this, and requires using filters to retain classes it might otherwise prune. I don't know how this feature works, but generally the idea would be to inspect the byte code to compute the class dependency graph. If any class is looked up by reflection then it would be lost and fail at runtime, like you observed. Can you try (1) disabling minimizeJar and if that works, then try (2) enabling with an appropriate filter?

Caffeine code generates the different cache and entry types in order to minimize the memory footprint. There are many configuration options which require different fields, so instead of paying for the worst case we use the optimized class forms. In most environments this is ideal because while it adds some bloat to the jar for deploying the artifact, unloaded classes have no runtime penalty. When the cache is built it uses reflection to load the particular implementation type. The reflective load is a one-time cost and cached by the JVM so that it is very cheap. However this means that ahead-of-time compilation tools must whitelist the appropriate classes as they cannot be statically determined.

**Reply by Andre601** [reply]
I tried removing `<minimizeJar>true</minimizeJar>` before, but only did so in the core and not also in the specific module. Removing it from both did seem to work now.

When I understand it correctly would I need to add a `<artifactSet>` with an `<includes>` list to the shading configuration, right?
Is there a recommended include pattern, or is `com.github.benmanes.caffeine` good enough?

**Reply by ben-manes** [reply]
I believe so, but I have not used Maven or its shade plugin in many years. You'll have to experiment.

The top-level package should be good enough. You can be more specific if you want to fully optimize to include only those classes that you use, but I don't think it is worth it on the JVM. Those using Graal AOT (like Quarkus) specify down to the exact classes because native binaries are much larger than the source jars.

**Reply by Andre601** [reply]
I played around with it and it seems I would now need to specific every dependency that needs to be included in order for all required stuff to be implemented. Otherwise Maven treats it as "If not listed, don't include".

Like right now I encounter the issues that classes from my main module (Meaning classes I myself made) are not included when I just set caffeine as an include with nothing else.

Another issue I see is, that Maven seems to include Caffeine without the relocation being made, which is bad as my use-case has a risk of other jars attempting to load the same dependencies to use, causing conflicts.

I decided to remove the `<minimizeJar>true</minimizeJar>` for now from the pom.xml files. The jars will be a lot larger, but that's a minor downside for getting things working.

**Reply by Andre601** [reply]
Actually, looking at the jar, it seems like that caffeine is included twice. Once as relocated content and once as its usual path... This is quite weird tbh... Need to look a bit further into this.

**Reply by ben-manes** [reply]
fwiw, you might need to use `<include>**</include>` per this [SO answer](https://stackoverflow.com/a/55183545/19450).

**Reply by Andre601** [reply]
~~I'm getting more and more confused with Maven...~~
~~As mentioned is it including a relocated and non-relocated version of Caffeine in the final jar. The interesting point is, that I do relocate it in the core module, which other modules implement as dependency, so it should be relocated at that point (Which it is), yet I have to declare a relocation in the other modules to relocate that com.github.benmanes path itself...~~

~~It's so confusing at times.~~

Managed to get rid of this by moving the relocation to the final modules, so that the core provides the stuff unshaded.
